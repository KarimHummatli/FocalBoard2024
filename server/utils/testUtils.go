package utils

import "github.com/stretchr/testify/mock"

var Anything = mock.MatchedBy(func(any) bool { return true })
