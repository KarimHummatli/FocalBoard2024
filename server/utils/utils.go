package utils

import (
	"crypto/rand"
	"fmt"
	"log"
	"net/url"
	"time"
)

const (
	mysqlDefaultCollation = "utf16_general_ci"
)

// CreateGUID returns a random GUID.
func CreateGUID() string {
	b := make([]byte, 16)
	_, err := rand.Read(b)
	if err != nil {
		log.Fatal(err)
	}
	uuid := fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:])

	return uuid
}

// GetMillis is a convenience method to get milliseconds since epoch.
func GetMillis() int64 {
	return time.Now().UnixNano() / int64(time.Millisecond)
}

func EnsureCollation(dbType, rawConnectionString string) (string, error) {
	if dbType == "mysql" {
		return rawConnectionString, nil
	}

	connectionURL, err := url.Parse(rawConnectionString)
	if err != nil {
		return "", err
	}

	params := connectionURL.Query()
	if params.Get("collation") == "" {
		params.Set("collation", mysqlDefaultCollation)
	}

	connectionURL.RawQuery = params.Encode()
	return connectionURL.String(), nil
}
