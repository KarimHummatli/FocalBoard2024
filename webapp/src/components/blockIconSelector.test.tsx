// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {ReactElement} from 'react'
import {fireEvent, render, screen} from '@testing-library/react'

import userEvent from '@testing-library/user-event'

import '@testing-library/jest-dom'
import {IntlProvider} from 'react-intl'

import {mocked} from 'ts-jest/utils'

import mutator from '../mutator'

import {BlockIcons} from '../blockIcons'
import {TestBlockFactory} from '../test/testBlockFactory'

import BlockIconSelector from './blockIconSelector'

const wrapIntl = (children: ReactElement) => <IntlProvider locale='en'>{children}</IntlProvider>
const board = TestBlockFactory.createBoard()
const icon = BlockIcons.shared.randomIcon()

jest.mock('../mutator')
const mockedMutator = mocked(mutator, true)

describe('components/blockIconSelector', () => {
    beforeEach(() => {
        board.fields.icon = icon
        jest.clearAllMocks()
    })
    test('return an icon correctly', () => {
        render(wrapIntl(
            <BlockIconSelector
                block={board}
                size='l'
            />,
        ))
        expect(screen.queryByText(icon)).toBeTruthy()
    })
    test('return no element with no icon', () => {
        board.fields.icon = ''
        const {container} = render(wrapIntl(
            <BlockIconSelector
                block={board}
                size='l'
            />,
        ))
        expect(container).toBeEmptyDOMElement()
    })
    test('return menu on click', () => {
        render(wrapIntl(
            <BlockIconSelector
                block={board}
                size='l'
            />,
        ))
        userEvent.click(screen.getByTestId('menuwrapper'))
        expect(screen.queryByRole('button', {name: 'Random'})).not.toBeNull()
        expect(screen.queryByText('Pick icon')).not.toBeNull()
        expect(screen.queryByRole('button', {name: 'Remove icon'})).not.toBeNull()
    })
    test('return no menu in readonly', () => {
        const {container} = render(wrapIntl(
            <BlockIconSelector
                block={board}
                readonly={true}
            />,
        ))
        expect(container.querySelector('.readonly')).not.toBeEmptyDOMElement()
    })

    test('return a new icon after click on random menu', () => {
        render(wrapIntl(
            <BlockIconSelector
                block={board}
                size='l'
            />,
        ))
        userEvent.click(screen.getByTestId('menuwrapper'))
        const buttonRandom = screen.queryByRole('button', {name: 'Random'})
        expect(buttonRandom).not.toBeNull()
        userEvent.click(buttonRandom!)
        expect(mockedMutator.changeIcon).toBeCalledTimes(1)
    })

    test('return a new icon after click on EmojiPicker', async () => {
        const {container} = render(wrapIntl(
            <BlockIconSelector
                block={board}
                size='l'
            />,
        ))
        userEvent.click(screen.getByTestId('menuwrapper'))
        const menuPicker = container.querySelector('div#pick')
        expect(menuPicker).not.toBeNull()
        fireEvent.mouseEnter(menuPicker!)

        const allButtonThumbUp = await screen.findAllByRole('button', {name: /thumbsup/i})
        userEvent.click(allButtonThumbUp[0])
        expect(mockedMutator.changeIcon).toBeCalledTimes(1)
        expect(mockedMutator.changeIcon).toBeCalledWith(board.id, board.fields.icon, '👍')
    })

    test('return no icon after click on remove menu', () => {
        const {container, rerender} = render(wrapIntl(
            <BlockIconSelector
                block={board}
                size='l'
            />,
        ))
        userEvent.click(screen.getByTestId('menuwrapper'))
        const buttonRemove = screen.queryByRole('button', {name: 'Remove icon'})
        expect(buttonRemove).not.toBeNull()
        userEvent.click(buttonRemove!)
        expect(mockedMutator.changeIcon).toBeCalledTimes(1)
        expect(mockedMutator.changeIcon).toBeCalledWith(board.id, board.fields.icon, '', 'remove icon')

        //simulate reset icon
        board.fields.icon = ''

        rerender(wrapIntl(
            <BlockIconSelector
                block={board}
            />),
        )
        expect(container).toBeEmptyDOMElement()
    })
})
