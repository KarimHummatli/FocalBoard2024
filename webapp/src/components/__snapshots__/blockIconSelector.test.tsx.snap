// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/blockIconSelector return an icon correctly 1`] = `
<div>
  <div
    class="IconSelector"
  >
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <div
        class="octo-icon size-l"
      >
        <span>
          üëç
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/blockIconSelector return menu on click 1`] = `
<div>
  <div
    class="IconSelector"
  >
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <div
        class="octo-icon size-l"
      >
        <span>
          üëç
        </span>
      </div>
      <div
        class="Menu noselect bottom "
      >
        <div
          class="menu-contents"
        >
          <div
            class="menu-options"
          >
            <div>
              <div
                aria-label="Random"
                class="MenuOption TextOption menu-option"
                role="button"
              >
                <div
                  class="d-flex"
                >
                  <div
                    class="menu-option__icon"
                  >
                    <svg
                      class="RandomIcon Icon"
                      viewBox="0 0 50 50"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M35.5 11.35H27C19.5559 11.35 13.5 17.4058 13.5 24.85C13.5 30.6388 8.78886 35.35 3 35.35H1C0.170858 35.35 -0.5 36.0208 -0.5 36.85C-0.5 37.6791 0.170858 38.35 1 38.35H3C10.4441 38.35 16.5 32.2941 16.5 24.85C16.5 19.0611 21.2111 14.35 27 14.35H35.5V22.814V23.7959L36.2941 23.2183L49.9941 13.2543L50.5501 12.85L49.9941 12.4456L36.2941 2.48162L35.5 1.90408V2.88599V11.35ZM38.5 17.9041V7.79584L45.4499 12.85L38.5 17.9041Z"
                      />
                      <path
                        d="M12.3368 14.4138L12.3371 14.414C13.1598 15.1506 13.2287 16.414 12.4941 17.2368M12.3368 14.4138L11.748 16.571M12.3368 14.4138C9.76805 12.1154 6.45152 10.85 3 10.85H1C-0.105285 10.85 -1 11.7447 -1 12.85C-1 13.9553 -0.105285 14.85 1 14.85H3C5.46754 14.85 7.83506 15.7537 9.66932 17.3944M12.3368 14.4138L10.336 16.649M12.4941 17.2368C12.494 17.2369 12.4939 17.237 12.4938 17.2372L11.748 16.571M12.4941 17.2368C12.4943 17.2367 12.4944 17.2365 12.4945 17.2364L11.748 16.571M12.4941 17.2368C12.0982 17.6808 11.5478 17.904 11.003 17.904C10.5302 17.904 10.0511 17.736 9.66932 17.3944M11.748 16.571C11.551 16.792 11.277 16.904 11.003 16.904C10.766 16.904 10.527 16.82 10.336 16.649M9.66932 17.3944L10.336 16.649M9.66932 17.3944C9.66921 17.3942 9.66909 17.3941 9.66898 17.394L10.336 16.649"
                      />
                      <path
                        d="M35.501 38.35V46.814V47.7959L36.2951 47.2183L49.9951 37.2543L50.5511 36.85L49.9951 36.4456L36.2941 26.4816L35.5 25.9041V26.886V35.35H27C24.4367 35.35 21.9697 34.4184 20.0534 32.7276C19.4346 32.1786 18.4852 32.2377 17.9361 32.8602L17.936 32.8603C17.3889 33.4808 17.4466 34.4287 18.0692 34.9769C20.5348 37.1527 23.7075 38.35 27.001 38.35H35.501ZM38.5 41.9041V31.7958L45.4499 36.85L38.5 41.9041Z"
                      />
                    </svg>
                  </div>
                </div>
                <div
                  class="menu-option__content"
                >
                  <div
                    class="menu-name"
                  >
                    Random
                  </div>
                </div>
                <div
                  class="noicon"
                />
              </div>
            </div>
            <div>
              <div
                class="MenuOption SubMenuOption menu-option"
                id="pick"
              >
                <div
                  class="menu-option__icon"
                >
                  <svg
                    class="EmojiIcon Icon"
                    viewBox="0 0 496 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 448c-110.3 0-200-89.7-200-200S137.7 56 248 56s200 89.7 200 200-89.7 200-200 200zm-80-216c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32zm160 0c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32zm4 72.6c-20.8 25-51.5 39.4-84 39.4s-63.2-14.3-84-39.4c-8.5-10.2-23.7-11.5-33.8-3.1-10.2 8.5-11.5 23.6-3.1 33.8 30 36 74.1 56.6 120.9 56.6s90.9-20.6 120.9-56.6c8.5-10.2 7.1-25.3-3.1-33.8-10.1-8.4-25.3-7.1-33.8 3.1z"
                    />
                  </svg>
                </div>
                <div
                  class="menu-name"
                >
                  Pick icon
                </div>
                <i
                  class="CompassIcon icon-chevron-right"
                />
              </div>
            </div>
            <div>
              <div
                aria-label="Remove icon"
                class="MenuOption TextOption menu-option"
                role="button"
              >
                <div
                  class="d-flex"
                >
                  <div
                    class="menu-option__icon"
                  >
                    <i
                      class="CompassIcon icon-trash-can-outline DeleteIcon trash-can-outline"
                    />
                  </div>
                </div>
                <div
                  class="menu-option__content"
                >
                  <div
                    class="menu-name"
                  >
                    Remove icon
                  </div>
                </div>
                <div
                  class="noicon"
                />
              </div>
            </div>
          </div>
          <div
            class="menu-spacer hideOnWidescreen"
          />
          <div
            class="menu-options hideOnWidescreen"
          >
            <div
              aria-label="Cancel"
              class="MenuOption TextOption menu-option menu-cancel"
              role="button"
            >
              <div
                class="d-flex"
              >
                <div
                  class="noicon"
                />
              </div>
              <div
                class="menu-option__content"
              >
                <div
                  class="menu-name"
                >
                  Cancel
                </div>
              </div>
              <div
                class="noicon"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/blockIconSelector return no element with no icon 1`] = `<div />`;

exports[`components/blockIconSelector return no icon after click on remove menu 1`] = `<div />`;

exports[`components/blockIconSelector return no menu in readonly 1`] = `
<div>
  <div
    class="IconSelector"
  >
    <div
      class="octo-icon size-m readonly"
    >
      <span>
        üëç
      </span>
    </div>
  </div>
</div>
`;
